external-crates/move/crates/move-trace-format/src/trace_format.rs

```diff
@@ -0,0 +1,387 @@
+    },
+    ImmRef {
+        location: Location,
+        snapshot: Box<MoveValue>,
+    },
+    MutRef {
+        location: Location,
+        snapshot: Box<MoveValue>,
+    },
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub enum RefType {
+    Imm,
+    Mut,
+}
+
+/// Type tag with references. This is a type tag that also supports references.
+/// if ref_type is None, this is a value type. If ref_type is Some, this is a reference type of the
+/// given reference type.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct TypeTagWithRefs {
+    pub type_: TypeTag,
+    pub ref_type: Option<RefType>,
+}
+
+/// A Frame represents a stack frame in the Move VM.
+/// This is an instantiation of `Function`.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct Frame {
+    // The frame id is the offset in the trace where this frame was opened.
+    pub frame_id: TraceIndex,
+    pub function_name: String,
+    pub module: ModuleId,
+    // External pointer out into the module
+    pub binary_member_index: u16,
+    pub type_instantiation: Vec<TypeTag>,
+    pub parameters: Vec<TraceValue>,
+    pub return_types: Vec<TypeTagWithRefs>,
+    pub locals_types: Vec<TypeTagWithRefs>,
+    pub is_native: bool,
+}
+
+/// An instruction effect is a single effect of an instruction. This can be a push/pop of a value,
+/// a read/write of a value, or a reference to a value.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub enum Effect {
+    // Pop a value off of the stack (pre-effect only)
+    Pop(TraceValue),
+    // Read a value from a location (pre-effect only)
+    Read(Read),
+
+    // Push a value off the stack (post-effect only)
+    Push(TraceValue),
+    // Write a value to a location (post-effect only)
+    Write(Write),
+
+    // A data load Effect
+    DataLoad(DataLoad),
+
+    // An execution error occured
+    ExecutionError(String),
+}
+
+/// Represent a data load event. This is a load of a value from storage. We only record loads by
+/// reference in the trace, and we snapshot the value at the reference location at the time of load
+/// and record its global reference ID (i.e., the location in the trace at which it was loaded).
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct DataLoad {
+    pub ref_type: RefType,
+    pub location: Location,
+    pub snapshot: MoveValue,
+}
+
+/// A TraceEvent is a single event in the Move VM, external events can also be interleaved in the
+/// trace. MoveVM events, are well structured, and can be a frame event or an instruction event.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
+pub enum TraceEvent {
+    OpenFrame {
+        frame: Frame,
+        gas_left: u64,
+    },
+    CloseFrame {
+        frame_id: TraceIndex,
+        return_: Vec<TraceValue>,
+        gas_left: u64,
+    },
+    Instruction {
+        type_parameters: Vec<TypeTag>,
+        pc: u16,
+        gas_left: u64,
+        instruction: String,
+    },
+    Effect(Effect),
```

Comment:
The main reason was multiple different events can have effects (but not always). Instructions have effects, but some close frame events do too -- if it's a native frame that is closed the returned values need to be `Push`ed on to the stack. It felt better to decouple effects into their own event type to provide better uniformity instead of having special cases in different places.